clients/java/signalr/src/test/java/com/microsoft/aspnet/signalr/HubConnectionTest.java

```diff
@@ -1053,6 +1055,64 @@ public void accessTokenProviderIsOverriddenFromRedirectNegotiate()
         assertEquals("Bearer newToken", token.get());
     }
 
+    @Test
+    public void headersAreSetAndSentThroughOptions() throws Exception {
+        AtomicReference<String> header = new AtomicReference<>();
+        TestHttpClient client = new TestHttpClient()
+                .on("POST", "http://example.com/negotiate",
+                        (req) -> {
+                            header.set(req.getHeaders().get("ExampleHeader"));
+                            return CompletableFuture
+                                    .completedFuture(new HttpResponse(200, "", "{\"connectionId\":\"bVOiRPG8-6YiJ6d7ZcTOVQ\",\""
+                                            + "availableTransports\":[{\"transport\":\"WebSockets\",\"transferFormats\":[\"Text\",\"Binary\"]}]}"));
+                        });
+
+        Map<String, String> headerMap = new HashMap<>();
+        headerMap.put("ExampleHeader", "ExampleValue");
+
+        MockTransport transport = new MockTransport();
+        HttpConnectionOptions options = new HttpConnectionOptions();
+        options.setTransport(transport);
+        options.setHttpClient(client);
+        options.setHeaders(headerMap);
+        HubConnection hubConnection = new HubConnectionBuilder()
+                .withUrl("http://example.com", options)
+                .build();
+
+        hubConnection.start().get(1000, TimeUnit.MILLISECONDS);
+        assertEquals(HubConnectionState.CONNECTED, hubConnection.getConnectionState());
+        hubConnection.stop();
+        assertEquals("ExampleValue", header.get());
+    }
+
+    @Test
+    public void headersAreSetAndSentThroughBuilder() throws Exception {
```

Comment:
Can you have a test that calls `withHeader` twice with the same value? I think the best thing to do is to have the second value win (which is what `put` should do by default).